### PR TITLE
More refactoring to take advantage of collective aggregators

### PR DESCRIPTION
**Pull request by rongou**
Also added `ShouldHaveLabels()` to encode when we expect to have labels.

### CODE REVIEW

File: src/collective/aggregator.h [hunk 1]
```diff
@@ -31,18 +31,16 @@ namespace collective {
  * @param buffer The buffer storing the results.
  * @param size The size of the buffer.
  * @param function The function used to calculate the results.
- * @param args Arguments to the function.
  */
-template <typename Function, typename T, typename... Args>
-void ApplyWithLabels(MetaInfo const& info, T* buffer, size_t size, Function&& function,
-                     Args&&... args) {
+template <typename Function>
+void ApplyWithLabels(MetaInfo const& info, void* buffer, size_t size, Function&& function) {
```
Comment (trivialfis) [inline src/collective/aggregator.h:36]: Is type erasion necessary? (void*)

Comment (rongou) [reply src/collective/aggregator.h:36]: When doing the broadcast, we don't need to know the type, only the size (https://github.com/dmlc/xgboost/blob/master/src/collective/communicator-inl.h#L128), so in here we don't really care about the type either.

**Comment by rongou** [comment]
@trivialfis @hcho3 Any ideas about the R test errors? Doesn't seem to be related to this PR.

**Comment by hcho3** [comment]
@rongou Please ignore it for now. We are aware of the issue (https://github.com/dmlc/xgboost/issues/9091). Not sure when I'd get time to fix it.

**Review by trivialfis — APPROVED** [review]

**Pull request merged**